src/ProductConstructionService/ProductConstructionService.Api/Queue/JobScope.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using ProductConstructionService.Api.Queue.JobRunners;
+using ProductConstructionService.Api.Queue.Jobs;
+using System.Diagnostics;
+using ProductConstructionService.Api.Metrics;
+
+namespace ProductConstructionService.Api.Queue;
+
+public class JobScope(
+    Action finalizer,
+    IServiceScope serviceScope) : IDisposable
+{
+    private readonly IServiceScope _serviceScope = serviceScope;
+    private Job? _job = null;
+
+    public void Dispose()
+    {
+        finalizer.Invoke();
+        _serviceScope.Dispose();
+    }
+
+    public void InitializeScope(Job job)
+    {
+        _job = job;
+    }
+
+    public async Task RunJob(CancellationToken cancellationToken)
+    {
+        if (_job is null)
+        {
+            throw new Exception("JobScope not initialized! Call InitializeScope before calling RunJob");
+        }
+
+        var jobRunner = _serviceScope.ServiceProvider.GetRequiredKeyedService<IJobRunner>(_job.GetType().Name);
+
+        var stopwatch = Stopwatch.StartNew();
+
+        await jobRunner.RunAsync(_job, cancellationToken);
+        
+        stopwatch.Stop();
+
+        var durationMetricRecorder = _serviceScope.ServiceProvider.GetRequiredService<MetricRecorder>();
+        durationMetricRecorder.RecordJobDuration(_job.GetType().Name, stopwatch.ElapsedMilliseconds);
```

Comment:
It might be cleaner if we just passed `_job` and the recorder would do with it whatever, e.g. log ID, name. But we shouldn't increase the complexity of the recorder API as it will be used from many places possibly.

Reply:
yeah let's pass the job, for the basic info